File: cpp/numerical-stacking-puzzle/celldat.cpp
/********************************************************************************
* Copyright 2026 Miyabi															*
*																				*
* Project Name	: Numerical Stacking Puzzle (Windows 11, Visual Studio 2022)	*
* File Name		: celldat.cpp													*
* Author		: Miyabi														*
* Create		: 2026/06/01 													*
* Purpose		: セルのデータ管理												*
********************************************************************************/

/********************************************************************************
*	Include Files																*
********************************************************************************/
#include "framework.h"	/* 必須 */
#include "define.h"
#include "celldat.h"

/********************************************************************************
*	Defines																		*
********************************************************************************/
// シーケンス
#define	E_SEQ_INIT	(0)	// 初期状態
#define	E_SEQ_IDLE	(1)	// 待機状態

/********************************************************************************
*	Define Static Instance														*
********************************************************************************/
CRCellDat*	CRCellDat::m_lpThis;

/********************************************************************************
*	CRCellDat Public Constructor(s)/Destructor									*
********************************************************************************/

/********************************************************************************
* FunctionName:	CRCellDat														*
* Function    :	コンストラクタ													*
* Argument(s) :	none															*
********************************************************************************/
CRCellDat::CRCellDat(void)
{
	//-----------------------------------------------------------------------
	// インスタンスポインタ保存
	m_lpThis = this;

	//-----------------------------------------------------------------------
	// 初期化処理
	// シーケンス設定（関数ポインタ配列作成）
	m_lpSeq[E_SEQ_INIT] = &CRCellDat::SeqInit;	// 初期状態
	m_lpSeq[E_SEQ_IDLE] = &CRCellDat::SeqIdle;	// 待機状態

	//-----------------------------------------------------------------------
	// バッファクリア
	memset(&(m_atatCellDat[0][0]), -1, sizeof(m_atatCellDat));

	//-----------------------------------------------------------------------
	// 初期値設定
	m_nScore = 0;			// 得点 <- 0
	m_nSeq = E_SEQ_INIT;	// 初期シーケンス=初期状態

	// 初期化正常終了
	this->CRoutine::InitOK();

	Debuglog((LPWSTR)(L"CRCellDat"));
}

/********************************************************************************
* FunctionName:	~CRCellDat														*
* Function    :	デストラクタ													*
* Argument(s) :	none															*
********************************************************************************/
CRCellDat::~CRCellDat(void)
{
	// インスタンスポインタ破棄
	m_lpThis = nullptr;

	Debuglog((LPWSTR)(L"~CRCellDat"));
}

/********************************************************************************
*	Private Methods																*
********************************************************************************/

/********************************************************************************
*	Private Methods (Sequence)													*
********************************************************************************/

/********************************************************************************
* FunctionName:	SeqInit															*
* Function    :	シーケンス処理:初期状態											*
* Argument(s) :	none															*
* Return      :	FALSE=処理終了（オブジェクト破棄）								*
********************************************************************************/
bool
CRCellDat::SeqInit(void)
{
	// シーケンス -> 待機状態
	m_nSeq = E_SEQ_IDLE;

	return TRUE;
}

/********************************************************************************
* FunctionName:	SeqIdle															*
* Function    :	シーケンス処理:待機状態											*
* Argument(s) :	none															*
* Return      :	FALSE=処理終了（オブジェクト破棄）								*
********************************************************************************/
bool
CRCellDat::SeqIdle(void)
{
	return TRUE;
}

/********************************************************************************
*	Public Methods																*
********************************************************************************/

/********************************************************************************
* FunctionName:	Update															*
* Function    :	定期更新処理													*
* Argument(s) :	none															*
* Return      :	FALSE=処理終了（オブジェクト破棄）								*
* Attention(s):	呼び出し周期＝1/60sec											*
********************************************************************************/
bool
CRCellDat::Update()
{
	//-----------------------------------------------------------------------
	// 基底クラスの処理
	if(this->CRoutine::Update() == FALSE)
	{
		return FALSE;
	}

	//-----------------------------------------------------------------------
	// シーケンス処理
	if((this->*m_lpSeq[m_nSeq])() == FALSE)
	{
		return FALSE;
	}

	return TRUE;
}

/********************************************************************************
* FunctionName:	SetVal															*
* Function    :	セル格納値を設定												*
* Argument    :	格納位置X,Y														*
*				格納値															*
* Return      :	none															*
********************************************************************************/
void
CRCellDat::SetVal(long nX, long nY, long nVal)
{
	m_atatCellDat[nX][nY].nVal = nVal;
}

/********************************************************************************
* FunctionName:	GetVal															*
* Function    :	セル格納値を取得												*
* Argument    :	格納位置X,Y														*
* Return      :	セル格納値														*
********************************************************************************/
long
CRCellDat::GetVal(long nX, long nY)
{
	return m_atatCellDat[nX][nY].nVal;
}

/********************************************************************************
* FunctionName:	GetStackVal														*
* Function    :	スタック情報を取得												*
* Argument    :	格納位置X,Y														*
* Return      :	スタック情報													*
********************************************************************************/
long
CRCellDat::GetStackVal(long nX, long nY)
{
	return m_atatCellDat[nX][nY].nStackVal;
}

/********************************************************************************
* FunctionName:	GetMergeVal														*
* Function    :	マージするセルの値を取得										*
* Argument    :	格納位置X,Y														*
* Return      :	マージするセルの値												*
********************************************************************************/
long
CRCellDat::GetMergeVal(long nX, long nY)
{
	return m_atatCellDat[nX][nY].nMergeVal;
}

/********************************************************************************
* FunctionName:	GetMergeX														*
* Function    :	マージ位置を取得												*
* Argument    :	格納位置X,Y														*
* Return      :	マージ位置														*
********************************************************************************/
long
CRCellDat::GetMergeX(long nX, long nY)
{
	return m_atatCellDat[nX][nY].nMergeX;
}

/********************************************************************************
* FunctionName:	GetMergeY														*
* Function    :	マージ位置を取得												*
* Argument    :	格納位置X,Y														*
* Return      :	マージ位置														*
********************************************************************************/
long
CRCellDat::GetMergeY(long nX, long nY)
{
	return m_atatCellDat[nX][nY].nMergeY;
}

/********************************************************************************
* FunctionName:	GetLandY														*
* Function    :	落下位置を取得													*
* Argument    :	格納位置X,Y														*
* Return      :	落下位置														*
********************************************************************************/
long
CRCellDat::GetLandY(long nX, long nY)
{
	return m_atatCellDat[nX][nY].nLandY;
}

/********************************************************************************
* FunctionName:	CheckEnStack													*
* Function    :	セル格納値の積み上げ可否を検査									*
* Argument    :	格納位置X,Y														*
*				格納値															*
* Return      :	積み上げ可否													*
********************************************************************************/
bool
CRCellDat::CheckEnStack(long nX, long nY, long nVal)
{
	long ret;				// 戻り値
	long nCntX, nCntY;		// 処理カウンタX,Y
	long nCellVal;			// セル側の格納値

	ret = FALSE;

	// セルデータのスタック情報を全消去
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			m_atatCellDat[nCntX][nCntY].nStackVal = -1;
		}
	}

	if(nY >= E_CELL_Y_NUM - 1)
	{
		m_atatCellDat[nX][nY].nStackVal = nVal;
		ret = TRUE;
	}
	else
	{
		nCellVal = m_atatCellDat[nX][nY+1].nVal;
		if(nCellVal >= 0)
		{
			if(nY >= 1)
			{
				m_atatCellDat[nX][nY].nStackVal = nVal;
				ret = TRUE;
			}
			else
			{
				if(nVal == nCellVal)
				{
					m_atatCellDat[nX][nY].nStackVal = nVal;
					ret = TRUE;
				}
			}
		}
	}

	return ret;
}

/********************************************************************************
* FunctionName:	ExecStack														*
* Function    :	セルデータの積み上げを実行										*
* Argument    :	none															*
* Return      :	none															*
********************************************************************************/
void
CRCellDat::ExecStack(void)
{
	long nCntX, nCntY;		// 処理カウンタX,Y

	// セルデータサイズ分繰り返す
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			if(m_atatCellDat[nCntX][nCntY].nStackVal >= 0)
			{
				m_atatCellDat[nCntX][nCntY].nVal = m_atatCellDat[nCntX][nCntY].nStackVal;
				m_atatCellDat[nCntX][nCntY].nStackVal = -1;
			}
		}
	}
}

/********************************************************************************
* FunctionName:	CheckMerge														*
* Function    :	セル格納値のマージ可否を検査									*
* Argument    :	格納位置X,Y														*
*				格納値															*
* Return      :	マージ可否														*
*				最大マージ値(ポインタ引数返し)									*
*				マージ規模(ポインタ引数返し)									*
********************************************************************************/
bool
CRCellDat::CheckMerge(long nX, long nY, long* lpMaxMergeVal, long* lpMaxMergeDifVal)
{
	bool ret;				// 戻り値
	long nCntX, nCntY;		// 処理カウンタX,Y
	long nMaxMergeVal;		// 最大マージ値
	long nMaxMergeDifVal;	// マージ規模

	ret = FALSE;

	// セルデータのマージ情報を全消去
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			m_atatCellDat[nCntX][nCntY].nMergeX = -1;
			m_atatCellDat[nCntX][nCntY].nMergeY = -1;
			m_atatCellDat[nCntX][nCntY].nMergeVal = -1;
		}
	}

	// セルデータサイズ分繰り返す(マージ可否検査)
	for(nCntX = nX; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			// 同値セルを再帰的に探索し、連結数をカウントする
			if(CheckMergeSub(nCntX, nCntY, nCntX, nCntY) > 1)
			{
				ret = TRUE;
			}
		}
	}
	for(nCntX = nX-1; nCntX >= 0; nCntX--)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			// 同値セルを再帰的に探索し、連結数をカウントする
			if(CheckMergeSub(nCntX, nCntY, nCntX, nCntY) > 1)
			{
				ret = TRUE;
			}
		}
	}

	// マージ可否検査結果をセット
	if(lpMaxMergeVal != nullptr)
	{
		nMaxMergeVal = -1;
		nMaxMergeDifVal = -1;
		if(ret != FALSE)
		{
			// セルデータサイズ分繰り返す
			for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
			{
				for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
				{
					if(m_atatCellDat[nCntX][nCntY].nVal < m_atatCellDat[nCntX][nCntY].nMergeVal)
					{
						nMaxMergeVal = Max(nMaxMergeVal, m_atatCellDat[nCntX][nCntY].nMergeVal);
						nMaxMergeDifVal = Max(nMaxMergeDifVal, m_atatCellDat[nCntX][nCntY].nMergeVal - m_atatCellDat[nCntX][nCntY].nVal);
					}
				}
			}
		}
		*lpMaxMergeVal = nMaxMergeVal;
		*lpMaxMergeDifVal = nMaxMergeDifVal;
	}
	
	return ret;
}

// ※privateでも良い
long
CRCellDat::CheckMergeSub(long nX, long nY, long nBaseX, long nBaseY)
{
	const long F_XY[4][2] = { {-1, 0},{ 1, 0},{ 0,-1},{ 0, 1} };
	long ret;
	long n;
	long nCellX;
	long nCellY;

	ret = 0;
	if(m_atatCellDat[nX][nY].nVal != -1 &&
	   m_atatCellDat[nX][nY].nMergeX == -1 &&
	   m_atatCellDat[nX][nY].nMergeY == -1)
	{
		if(m_atatCellDat[nX][nY].nVal == m_atatCellDat[nBaseX][nBaseY].nVal)
		{
			ret = 1;
			m_atatCellDat[nX][nY].nMergeX = nBaseX;
			m_atatCellDat[nX][nY].nMergeY = nBaseY;
			for(n = 0; n < 4; n++)
			{
				nCellX = nX + F_XY[n][0];
				nCellY = nY + F_XY[n][1];
				if(NUMOK(nCellX, E_CELL_X_NUM) &&
				   NUMOK(nCellY, E_CELL_Y_NUM))
				{
					ret += CheckMergeSub(nCellX, nCellY, nBaseX, nBaseY);
				}
			}

			if(nX == nBaseX &&
			   nY == nBaseY)
			{
				m_atatCellDat[nX][nY].nMergeVal = m_atatCellDat[nX][nY].nVal + ret - 1;
			}
		}
	}

	return ret;
}

/********************************************************************************
* FunctionName:	ExecMerge														*
* Function    :	セル格納値のマージを実行										*
* Argument    :	none															*
* Return      :	none															*
********************************************************************************/
void
CRCellDat::ExecMerge(void)
{
	long nCntX;	// 処理カウンタX
	long nCntY;	// 処理カウンタY

	// セルデータサイズ分繰り返す
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			m_atatCellDat[nCntX][nCntY].nVal = m_atatCellDat[nCntX][nCntY].nMergeVal;
			m_atatCellDat[nCntX][nCntY].nMergeX = -1;
			m_atatCellDat[nCntX][nCntY].nMergeY = -1;
			m_atatCellDat[nCntX][nCntY].nMergeVal = -1;
		}
	}
}

/********************************************************************************
* FunctionName:	CheckEnLand														*
* Function    :	セル格納値の着地可否を検査										*
* Argument    :	none															*
* Return      :	着地可否														*
********************************************************************************/
bool
CRCellDat::CheckEnLand(void)
{
	bool ret;				// 戻り値
	long nCntX, nCntY;		// 処理カウンタX,Y

	ret = FALSE;

	// セルデータの落下情報を全消去
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			m_atatCellDat[nCntX][nCntY].nLandY = -1;
		}
	}

	// セルデータサイズ分繰り返す
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = E_CELL_Y_NUM - 1; nCntY >= 0; nCntY--)
		{
			// 落下可能位置を計算
			if(CheckEnLandSub(nCntX, nCntY) != FALSE)
			{
				ret = TRUE;
			}
		}
	}

	return ret;
}

// ※privateでも良い
bool
CRCellDat::CheckEnLandSub(long nX, long nY)
{
	bool ret;
	long nCellY;

	ret = FALSE;
	if(m_atatCellDat[nX][nY].nVal != -1)
	{
		if(nY < E_CELL_Y_NUM - 1)
		{
			for(nCellY = nY + 1; nCellY < E_CELL_Y_NUM; nCellY++)
			{
				if(m_atatCellDat[nX][nCellY].nVal == -1)
				{
					m_atatCellDat[nX][nY].nLandY = nCellY;
					ret = TRUE;
				}
				else
				{
					if(m_atatCellDat[nX][nCellY].nLandY >= 0)
					{
						m_atatCellDat[nX][nY].nLandY = m_atatCellDat[nX][nCellY].nLandY - 1;
						ret = TRUE;
					}
					break;
				}
			}
		}
	}

	return ret;
}

/********************************************************************************
* FunctionName:	ExecLand														*
* Function    :	セル格納値の着地を実行											*
* Argument    :	none															*
* Return      :	none															*
********************************************************************************/
void
CRCellDat::ExecLand(void)
{
	long nCntX, nCntY;		// 処理カウンタX,Y
	long nLandY;			// 落下位置Y

	// セルデータサイズ分繰り返す
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = E_CELL_Y_NUM-1; nCntY >= 0; nCntY--)
		{
			nLandY = m_atatCellDat[nCntX][nCntY].nLandY;
			if(nLandY != -1)
			{
				m_atatCellDat[nCntX][nLandY].nVal = m_atatCellDat[nCntX][nCntY].nVal;
				m_atatCellDat[nCntX][nCntY].nVal = -1;
			}
			m_atatCellDat[nCntX][nCntY].nLandY = -1;
		}
	}
}

/********************************************************************************
* FunctionName:	CalcScore														*
* Function    :	得点計算														*
* Argument    :	none															*
* Return      :	得点増加の有無													*
********************************************************************************/
bool
CRCellDat::CalcScore(void)
{
	bool ret;				// 戻り値
	long nCntX, nCntY;		// 処理カウンタX,Y
	long nScore;			// スコア
	long nVal;				// セル格納値

	ret = FALSE;

	nScore = 0;

	// セルデータの数値を合算する
	for(nCntX = 0; nCntX < E_CELL_X_NUM; nCntX++)
	{
		for(nCntY = 0; nCntY < E_CELL_Y_NUM; nCntY++)
		{
			nVal = m_atatCellDat[nCntX][nCntY].nVal;
			if(nVal >= 0 && nVal <= 30)
			{
				nScore += (1<<nVal);
			}
			if(nScore < 0)
			{
				// オーバーフロー
				break;
			}
		}
		if(nScore < 0)
		{
			// オーバーフロー
			break;
		}
	}

	if(nScore >= 0)
	{
		m_nScore = nScore;
		ret = TRUE;
	}
	
	return ret;
}

/********************************************************************************
* FunctionName:	GetScore														*
* Function    :	得点を取得														*
* Argument    :	none															*
* Return      :	得点															*
********************************************************************************/
long
CRCellDat::GetScore(void)
{
	return m_nScore;
}

/********************************************************************************
* FunctionName:	SetScore														*
* Function    :	得点を設定														*
* Argument    :	得点															*
* Return      :	none															*
********************************************************************************/
void
CRCellDat::SetScore(long nScore)
{
	m_nScore = nScore;
}

File: python/numerical-stacking-puzzle/celldat.py
"""
Copyright 2026 Miyabi
Project Name  : Numerical Stacking Puzzle (Python)
File Name     : celldat.py
Author        : Miyabi
Create        : 2026/06/01
Purpose       : セルのデータ管理
"""

#
# Import
#
from __future__ import annotations
from routine import Routine
from define import E_CELL_X_NUM, E_CELL_Y_NUM

class Cell:
    def __init__(self):
        self.nVal = -1
        self.nStackVal = -1
        self.nMergeVal = -1
        self.nMergeX = -1
        self.nMergeY = -1
        self.nLandY = -1
        self.nLandVal = -1

#
# Definition of Class
#
class CellDat(Routine):
    _inst = None

    @staticmethod
    def get_inst():
        return CellDat._inst

    #----- Constructor(s)/Destructor
    def __init__(self):
        """コンストラクタ
        """
        super().__init__()

        #------------------------------------------------------------------------
        # インスタンスポインタ保存
        CellDat._inst = self

        # バッファクリア
        self.cell = [
            [Cell() for _ in range(E_CELL_Y_NUM)]
            for _ in range(E_CELL_X_NUM)
        ]

        #------------------------------------------------------------------------
        # 初期値設定
        self.score = 0      # 得点 <- 0

        # 初期化正常終了
        self.init_ok_mark()

    def set_val(self, x, y, v):
        """セル格納値を設定

        Args:
            格納位置X,Y
            格納値
        """
        self.cell[x][y].nVal = int(v)

    def get_val(self, x, y):
        """セル格納値を取得

        Args:
            格納位置X,Y

        Returns:
            セル格納値
        """
        return self.cell[x][y].nVal

    def get_stack_val(self, x, y):
        """スタック情報を取得

        Args:
            格納位置X,Y

        Returns:
            スタック情報
        """
        return self.cell[x][y].nStackVal

    def get_merge_val(self, x, y):
        """マージするセルの値を取得

        Args:
            格納位置X,Y

        Returns:
            マージするセルの値
        """
        return self.cell[x][y].nMergeVal

    def get_merge_x(self, x, y):
        """マージ位置を取得

        Args:
            格納位置X,Y

        Returns:
            マージ位置
        """
        return self.cell[x][y].nMergeX

    def get_merge_y(self, x, y):
        """マージ位置を取得

        Args:
            格納位置X,Y

        Returns:
            マージ位置
        """
        return self.cell[x][y].nMergeY

    def get_land_y(self, x, y):
        """落下位置を取得

        Args:
            格納位置X,Y

        Returns:
            落下位置
        """

        return self.cell[x][y].nLandY

    def get_score(self):
        """得点を取得

        Returns:
            得点
        """
        return self.score

    def set_score(self, s):
        """得点を設定

        Args:
            得点
        """
        self.score = int(s)

    def check_en_stack(self, x, y, val):
        """セル格納値の積み上げ可否を検査

        Args:
            格納位置X,Y
            格納値

        Returns:
            積み上げ可否
        """
        # セルデータのスタック情報を全消去
        for ix in range(E_CELL_X_NUM):
            for iy in range(E_CELL_Y_NUM):
                self.cell[ix][iy].nStackVal = -1

        if y >= E_CELL_Y_NUM - 1:
            self.cell[x][y].nStackVal = val
            return True

        below = self.cell[x][y + 1].nVal
        if below >= 0:
            if y >= 1:
                self.cell[x][y].nStackVal = val
                return True
            else:
                if val == below:
                    self.cell[x][y].nStackVal = val
                    return True
        return False

    def exec_stack(self):
        """セルデータの積み上げを実行
        """
        for x in range(E_CELL_X_NUM):
            for y in range(E_CELL_Y_NUM):
                c = self.cell[x][y]
                if c.nStackVal >= 0:
                    c.nVal = c.nStackVal
                    c.nStackVal = -1

    def check_merge(self, base_x, base_y):
        """セル格納値のマージ可否を検査

        Args:
            格納位置X,Y
            格納値

        Return:マージ可否
            最大マージ値(ポインタ引数返し)
            マージ規模(ポインタ引数返し)
        """
        # セルデータのマージ情報を全消去
        for x in range(E_CELL_X_NUM):
            for y in range(E_CELL_Y_NUM):
                c = self.cell[x][y]
                c.nMergeX = -1
                c.nMergeY = -1
                c.nMergeVal = -1

        found = False

        # セルデータサイズ分繰り返す(マージ可否検査)
        for x in range(base_x, E_CELL_X_NUM):
            for y in range(E_CELL_Y_NUM):
                # 同値セルを再帰的に探索し、連結数をカウントする
                if self._merge_sub(x, y, x, y) > 1:
                    found = True
        for x in range(base_x - 1, -1, -1):
            for y in range(E_CELL_Y_NUM):
                # 同値セルを再帰的に探索し、連結数をカウントする
                if self._merge_sub(x, y, x, y) > 1:
                    found = True

        # マージ可否検査結果をセット
        max_val = -1
        max_diff = -1

        if found:
            # セルデータサイズ分繰り返す
            for x in range(E_CELL_X_NUM):
                for y in range(E_CELL_Y_NUM):
                    c = self.cell[x][y]
                    if c.nVal < c.nMergeVal:
                        if c.nMergeVal > max_val:
                            max_val = c.nMergeVal
                        diff = c.nMergeVal - c.nVal
                        if diff > max_diff:
                            max_diff = diff

        return found, max_val, max_diff

    def _merge_sub(self, x, y, bx, by):
        dirs = [(-1,0),(1,0),(0,-1),(0,1)]
        c = self.cell[x][y]

        if c.nVal == -1:
            return 0

        if c.nMergeX != -1:
            return 0

        if c.nVal != self.cell[bx][by].nVal:
            return 0

        c.nMergeX = bx
        c.nMergeY = by
        count = 1

        for dx, dy in dirs:
            nx, ny = x + dx, y + dy
            if 0 <= nx < E_CELL_X_NUM and 0 <= ny < E_CELL_Y_NUM:
                count += self._merge_sub(nx, ny, bx, by)

        if x == bx and y == by:
            c.nMergeVal = c.nVal + count - 1

        return count

    def exec_merge(self):
        """セル格納値のマージを実行
        """
        # セルデータサイズ分繰り返す
        for x in range(E_CELL_X_NUM):
            for y in range(E_CELL_Y_NUM):
                c = self.cell[x][y]
                c.nVal = c.nMergeVal
                c.nMergeX = -1
                c.nMergeY = -1
                c.nMergeVal = -1

    def check_en_land(self):
        """セル格納値の着地可否を検査

        Return:
            着地可否
        """
        ret = False

        # セルデータの落下情報を全消去
        for x in range(E_CELL_X_NUM):
            for y in range(E_CELL_Y_NUM):
                self.cell[x][y].nLandY = -1

        # セルデータサイズ分繰り返す
        for x in range(E_CELL_X_NUM):
            for y in reversed(range(E_CELL_Y_NUM)):
                # 落下可能位置を計算
                if self._land_sub(x, y):
                    ret = True

        return ret

    def _land_sub(self, x, y):
        c = self.cell[x][y]
        if c.nVal == -1:
            return False

        if y >= E_CELL_Y_NUM - 1:
            return False

        for oy in range(y + 1, E_CELL_Y_NUM):
            if self.cell[x][oy].nVal == -1:
                c.nLandY = oy
                return True
            elif self.cell[x][oy].nLandY >= 0:
                c.nLandY = self.cell[x][oy].nLandY - 1
                return True
            else:
                break

        return False

    def exec_land(self):
        """セル格納値の着地を実行
        """
        # セルデータサイズ分繰り返す
        for x in range(E_CELL_X_NUM):
            for y in reversed(range(E_CELL_Y_NUM)):
                land_y = self.cell[x][y].nLandY
                if land_y != -1:
                    self.cell[x][land_y].nVal = self.cell[x][y].nVal
                    self.cell[x][y].nVal = -1
                self.cell[x][y].nLandY = -1

    def calc_score(self):
        """得点計算

        Returns:
            得点増加の有無
        """
        score = 0
        # セルデータの数値を合算する
        for x in range(E_CELL_X_NUM):
            for y in range(E_CELL_Y_NUM):
                v = self.cell[x][y].nVal
                if 0 <= v <= 30:
                    score += (1 << v)
                if score < 0:
                    # オーバーフロー
                    return False

        self.score = score
        return True

    def update(self):
        """定期更新処理

        Returna:
            False=処理終了（オブジェクト破棄）

        Notes:
            呼び出し周期＝1/60sec
        """
        return True
